### PR TITLE
Expand unboxing for Nullable<> in JIT

### LINKED ISSUE #104954: Optimize unboxing for nullable
```cs
int? Unbox(object o) => (int?)o;
```
Currently generates:
```asm
; Method Prog:Foo(System.Object):System.Nullable`1[int] (FullOpts)
G_M28087_IG01:
       sub      rsp, 40
       mov      r8, rcx
G_M28087_IG02:
       lea      rcx, [rsp+0x20]
       mov      rdx, 0x7FF8318BB750      ; System.Nullable`1[int]
       call     CORINFO_HELP_UNBOX_NULLABLE
       mov      rax, qword ptr [rsp+0x20]
G_M28087_IG03:
       add      rsp, 40
       ret      
; Total bytes of code: 37
```
we should be able to avoid the unbox helper call here by inlining the fast path. E.g.

```cs
Nullable<int> stackLocal;
if (obj == null)
{
    stackLocal = {0};
}
else if (obj->pMT == int)
{
    stackLocal._hasValue = 1;
    stackLocal._value = *((int*)(obj + 8);
}
else
{
    CORINFO_HELP_UNBOX_NULLABLE(&stackLocal, obj, clsHandle); // cold fallback (no return?)
}
```
NOTE: no, we can't unbox `int?` out of boxed `uint` - it's InvalidCastException so we're fine here.

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/104954

Currently, unboxing for Nullable<> always go through a helper call, while we can inline it in JIT.
```cs
int? Test(object o) => (int?)o;
```
#### Current codegen:
```asm
; Method Proga:Test(System.Object):System.Nullable`1[int]:this (FullOpts)
G_M11066_IG01:
       sub      rsp, 40
G_M11066_IG02:
       lea      rcx, [rsp+0x20]
       mov      r8, rdx
       mov      rdx, 0x7FF7E261B750      ; System.Nullable`1[int]
       call     CORINFO_HELP_UNBOX_NULLABLE
       mov      rax, qword ptr [rsp+0x20]
G_M11066_IG03:
       add      rsp, 40
       ret      
; Total bytes of code: 37
```
#### New codegen:
```asm
; Method Proga:Test(System.Object):System.Nullable`1[int]:this (FullOpts)
G_M11066_IG01:
       sub      rsp, 40
G_M11066_IG02:
       test     rdx, rdx ;; is null?
       je       SHORT G_M11066_IG04
G_M11066_IG03:
       mov      rax, 0x7FF7E23674D0      ; System.Int32
       cmp      qword ptr [rdx], rax ;; is object's type int32?
       je       SHORT G_M11066_IG05
       jmp      SHORT G_M11066_IG08
G_M11066_IG04:
       xor      edx, edx
       mov      qword ptr [rsp+0x20], rdx
       jmp      SHORT G_M11066_IG06
G_M11066_IG05:
       mov      byte  ptr [rsp+0x20], 1
       mov      eax, dword ptr [rdx+0x08]
       mov      dword ptr [rsp+0x24], eax
G_M11066_IG06:
       mov      rax, qword ptr [rsp+0x20]
G_M11066_IG07:
       add      rsp, 40
       ret      

; cold path
G_M11066_IG08:
       lea      rcx, [rsp+0x20]
       mov      r8, rdx
       mov      rdx, 0x7FF7E263B750      ; System.Nullable`1[int]
       call     CORINFO_HELP_UNBOX_NULLABLE
       jmp      SHORT G_M11066_IG06
; Total bytes of code: 82
```

#### Benchmark:
```cs
using BenchmarkDotNet.Attributes;

public class MyBench
{
    [Benchmark]
    [Arguments(42)]
    [Arguments(null)]
    public int? Unbox(object o) => (int?)o;
}
```
```
BenchmarkDotNet v0.13.12, Ubuntu 22.04.4 LTS (Jammy Jellyfish)
Unknown processor (which is "Ampere Altra")
  Job-MMNLPR : .NET 9.0.0 (42.42.42.42424), Arm64 RyuJIT AdvSIMD
  Job-QXOQTU : .NET 9.0.0 (42.42.42.42424), Arm64 RyuJIT AdvSIMD
```
| Method | Toolchain               | o  | Mean       | Error     | Ratio |
|------- |------------------------ |--- |-----------:|----------:|------:|
| **Unbox**  | **Main** | **?**  |  **7.1195 ns** | **0.0098 ns** |  **1.00** |
| Unbox  | PR | ?  |  0.6698 ns | 0.0072 ns |  0.09 |
|        |                         |    |            |           |       |
| **Unbox**  | **Main** | **42** | **10.9406 ns** | **0.0175 ns** |  **1.00** |
| Unbox  | PR | 42 |  2.7002 ns | 0.0001 ns |  0.25 |

### CODE REVIEW

**Comment by EgorBo** [comment]
@EgorBot -intel -arm64
```cs
using BenchmarkDotNet.Attributes;

public class MyBench
{
    [Benchmark]
    [Arguments(42)]
    [Arguments(null)]
    public int? Unbox(object o) => (int?)o;
}
```

**Comment by huoyaoyuan** [comment]
Does it worth to recognize `RuntimeHelpers.Unbox_Nullable` and do the same expansion?

Or, can we write the method in C# and make helpers inlinable? This will help more around nullables.

**Comment by EgorBo** [comment]
> Or, can we write the method in C# and make helpers inlinable? This will help more around nullables.

Yes, that is an option, requires some effort to make helpers inlineable, they're currently not. I decided to do it in JIT since the actual change is like 30 LOC, the rest are comments. If we move it to an inlineable helper we can delete this I guess..

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@jakobbotsch @AndyAyersMS cc @dotnet/jit-contrib PTAL, the actual change is 40 LOC, benchmarks attached. Jit-diff: https://github.com/MihuBot/runtime-utils/issues/541 obviously, a size regression, but there were also a few cases where jit can now fold the `if` we emit here, e.g. previously we never tried to fold `CORINFO_HELP_UNBOX_NULLABLE(null)` etc.

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -2832,6 +2832,96 @@ GenTree* Compiler::impImportLdvirtftn(GenTree*                thisPtr,
...
+    CORINFO_CLASS_HANDLE unboxType = info.compCompHnd->getTypeForBox(nullableCls);
+
+    // Clone the object (and spill side effects)
+    GenTree* objClone;
+    obj = impCloneExpr(obj, &objClone, CHECK_SPILL_ALL, nullptr DEBUGARG("op1 spilled for Nullable unbox"));
+
+    // Unbox the object to the result local:
+    //
+    //  result._hasValue = true;
+    //  result._value = MethodTableLookup(obj);
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:2879]: oops, wrong comment, it's `obj->RawData`

**Review by AndyAyersMS — APPROVED** [review]
Would it be worth limiting this to relatively small types, since will create both a zeroing and a copy?

**Comment by EgorBo** [comment]
Might be good idea to limit it, yeah

@EgorBot -intel -arm64
```cs
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

public class MyBench2
{
    [InlineArray(100)]
    public struct LargeStruct
    {
        public long Fld;
    }

    [Benchmark]
    [Arguments(42)]
    [Arguments(null)]
    public LargeStruct? Unbox(object o) => (MyStruct?)o;
}
```

**Comment by EgorBo** [comment]
Might be good idea to limit it, yeah

@EgorBot -intel -arm64
```cs
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

public class MyBench2
{
    [InlineArray(100)]
    public struct LargeStruct
    {
        public long Fld;
    }

    [Benchmark]
    [Arguments(42)]
    [Arguments(null)]
    public LargeStruct? Unbox(object o) => (LargeStruct?)o;
}
```

**Comment by EgorBo** [comment]
@EgorBot -intel -arm64
```cs
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

public class MyBench2
{
    [InlineArray(100)]
    public struct LargeStruct
    {
        public long Fld;
    }

    public IEnumerable<object> TestData()
    {
        yield return null;
        yield return new LargeStruct();
    }

    [Benchmark]
    [ArgumentsSource(nameof(TestData))]
    public LargeStruct? Unbox(object o) => (LargeStruct?)o;
}
```

**Pull request merged**